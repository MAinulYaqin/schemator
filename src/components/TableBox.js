import React, { forwardRef } from "react";
import PropTypes from "prop-types";
import styled from "styled-components";

import "typeface-roboto";

import TableHeader from "./TableHeader";
import TableButton from "./TableButton";
import TableInput from "./TableInput";
import TableOption from "./TableOption";

const Container = styled.div`
  width: 240px;
  background: #222;
  box-shadow: 0px 0px 10px rgba(0, 0, 0, 0.5);
  box-sizing: border-box;
  border: 2px solid #111;
`;

const TableBox = forwardRef(
  (
    {
      position,
      name,
      fields,
      options,
      onChangeName,
      onChangeFieldName,
      onChangeFieldType,
      onChangeOptions,
      onClickAddField,
      onClickRemoveField,
      onMouseDown,
      onMouseMove,
      onMouseEnter,
      onMouseLeave
    },
    ref
  ) => {
    const { x, y } = position;
    const headerHeight = 36;
    const buttonHeight = 36;
    const fieldHeight = 36 * fields.length;
    const optionHeight = 72;
    const height = headerHeight + buttonHeight + fieldHeight + optionHeight + 4;

    return (
      <foreignObject
        ref={ref}
        x={x}
        y={y}
        width={240}
        height={height}
        onMouseDown={onMouseDown}
        onMouseMove={onMouseMove}
        onMouseEnter={onMouseEnter}
        onMouseLeave={onMouseLeave}
      >
        <Container>
          <TableHeader caption={name} onChangeCaption={onChangeName} />
          {fields.map(field => (
            <TableInput
              key={field.id}
              name={field.name}
              type={field.type}
              onChangeName={event => onChangeFieldName(event, field.id)}
              onChangeType={event => onChangeFieldType(event, field.id)}
              onClickRemove={() => onClickRemoveField(field.id)}
            />
          ))}
          <TableOption onChange={onChangeOptions} value={options} />
          <TableButton caption="Add New Field" onClick={onClickAddField} />
        </Container>
      </foreignObject>
    );
  }
);

TableBox.propTypes = {
  position: PropTypes.object,
  name: PropTypes.string,
  fields: PropTypes.array,
  options: PropTypes.object,
  onMouseDown: PropTypes.func,
  onMouseMove: PropTypes.func,
  onMouseEnter: PropTypes.func,
  onMouseLeave: PropTypes.func,
  onClickAddField: PropTypes.func,
  onClickRemoveField: PropTypes.func,
  onChangeFieldName: PropTypes.func,
  onChangeFieldType: PropTypes.func,
  onChangeName: PropTypes.func,
  onChangeOptions: PropTypes.func
};

export default TableBox;
