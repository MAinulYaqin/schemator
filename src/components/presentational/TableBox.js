import React, { forwardRef } from "react";
import PropTypes from "prop-types";
import styled from "styled-components";

import "typeface-roboto";

import TableHeader from "./TableHeader";
import TableButton from "./TableButton";
import TableInput from "./TableInput";
import TableOption from "./TableOption";

const Container = styled.div`
  width: 240px;
  background: #222;
  box-shadow: 0px 0px 10px rgba(0, 0, 0, 0.5);
  box-sizing: border-box;
  border: 2px solid #111;
`;

const TableBox = forwardRef(
  (
    {
      position,
      types,
      name,
      fields,
      options,
      onChangeName,
      onChangeFieldName,
      onChangeFieldType,
      onChangeOptions,
      onClickAddField,
      onClickRemoveField,
      onMouseDown,
      onMouseUp,
      onMouseMove,
      onMouseEnter,
      onMouseLeave,
      onContextMenu
    },
    ref
  ) => {
    const { x, y } = position;
    const headerHeight = 36;
    const buttonHeight = 36;
    const fieldHeight = 36 * fields.length;
    const optionHeight = 72;
    const height = headerHeight + buttonHeight + fieldHeight + optionHeight + 4;

    // Count total needed of TableOption
    const totalOptionChunks = Math.round(options.length / 2);

    // Split options into several chunks.
    // Each chunk contains 2 options.
    const optionChunks = [...Array(totalOptionChunks)].map((_, index) =>
      options.slice(index * 2, (index + 1) * 2)
    );

    return (
      <foreignObject
        ref={ref}
        x={x}
        y={y}
        width={240}
        height={height}
        onMouseDown={onMouseDown}
        onMouseUp={onMouseUp}
        onMouseMove={onMouseMove}
        onMouseEnter={onMouseEnter}
        onMouseLeave={onMouseLeave}
        onContextMenu={onContextMenu}
      >
        <Container>
          <TableHeader caption={name} onChangeCaption={onChangeName} />
          {fields.map(field => (
            <TableInput
              key={field.id}
              types={types}
              name={field.name}
              type={field.type}
              onChangeName={event => onChangeFieldName(event, field.id)}
              onChangeType={event => onChangeFieldType(event, field.id)}
              onClickRemove={() => onClickRemoveField(field.id)}
            />
          ))}
          {optionChunks.map((chunk, index) => (
            <TableOption key={index} onChange={onChangeOptions} items={chunk} />
          ))}
          <TableButton caption="Add New Field" onClick={onClickAddField} />
        </Container>
      </foreignObject>
    );
  }
);

TableBox.propTypes = {
  position: PropTypes.object,
  name: PropTypes.string,
  fields: PropTypes.array,
  options: PropTypes.array,
  onMouseDown: PropTypes.func,
  onMouseUp: PropTypes.func,
  onMouseMove: PropTypes.func,
  onMouseEnter: PropTypes.func,
  onMouseLeave: PropTypes.func,
  onContextMenu: PropTypes.func,
  onClickAddField: PropTypes.func,
  onClickRemoveField: PropTypes.func,
  onChangeFieldName: PropTypes.func,
  onChangeFieldType: PropTypes.func,
  onChangeName: PropTypes.func,
  onChangeOptions: PropTypes.func
};

export default TableBox;
