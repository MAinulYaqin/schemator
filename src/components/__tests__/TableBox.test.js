import React from "react";
import { shallow } from "enzyme";
import toJSON from "enzyme-to-json";

import TableBox from "../TableBox";

describe("TableBox", () => {
  const setup = propOverrides => {
    const props = {
      position: { x: 32, y: 32 },
      name: "User",
      fields: [
        {
          id: "1",
          name: "id",
          type: "INCREMENT"
        },
        {
          id: "2",
          name: "username",
          type: "VARCHAR"
        }
      ],
      onClickAddField: jest.fn(),
      onClickRemoveField: jest.fn(),
      onChangeFieldName: jest.fn(),
      onChangeFieldType: jest.fn(),
      onChangeName: jest.fn(),
      onMouseDown: jest.fn(),
      onMouseMove: jest.fn(),
      ...propOverrides
    };

    const wrapper = shallow(<TableBox {...props} />);

    return {
      wrapper,
      props
    };
  };

  it("should renders default", () => {
    const { wrapper } = setup();

    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  it("should calls 'onMouseDown' while mouse down", () => {
    const { wrapper, props } = setup();

    wrapper.simulate("mouseDown");

    expect(props.onMouseDown).toBeCalled();
  });

  it("should calls 'onMouseMove' while mouse move", () => {
    const { wrapper, props } = setup();

    wrapper.simulate("mouseMove");

    expect(props.onMouseMove).toBeCalled();
  });
});
