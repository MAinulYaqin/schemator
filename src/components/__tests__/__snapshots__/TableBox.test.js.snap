// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableBox should renders default 1`] = `
<foreignObject
  height={148}
  onMouseDown={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onMouseMove={[Function]}
  width={240}
  x={32}
  y={32}
>
  <styled.div>
    <TableHeader
      caption="User"
      onChangeCaption={[Function]}
    />
    <TableInput
      key="1"
      name="id"
      onChangeName={[Function]}
      onChangeType={[Function]}
      onClickRemove={[Function]}
      type="INCREMENT"
    />
    <TableInput
      key="2"
      name="username"
      onChangeName={[Function]}
      onChangeType={[Function]}
      onClickRemove={[Function]}
      type="VARCHAR"
    />
    <TableButton
      caption="Add New Field"
      onClick={[Function]}
    />
  </styled.div>
</foreignObject>
`;
